Main.hs

```diff
@@ -18,29 +18,99 @@ import qualified Network.Wai as Wai
 import qualified Network.Wai.Handler.Warp as Warp
 import           Network.Wai.Handler.WebSockets (websocketsOr)
 import           Network.WebSockets.Connection (defaultConnectionOptions)
+import           Routes
 import qualified Sockets
 import           System.Random (getStdRandom, randomR)
 import           Web.Scotty
 
-
 maybeParam :: Parsable a => Lazy.Text -> ActionM (Maybe a)
 maybeParam name = (Just <$> param name) `rescue` (return . const Nothing)
 
-getHeader :: HeaderName -> ActionM (Maybe Strict.Text)
-getHeader name = (listToMaybe . map (Strict.decodeUtf8 . snd)
-                  . filter ((== name) . fst) . Wai.requestHeaders) <$> request
+maybeHeader :: HeaderName -> ActionM (Maybe Strict.Text)
+maybeHeader name = (listToMaybe . map (Strict.decodeUtf8 . snd)
+                   . filter ((== name) . fst) . Wai.requestHeaders) <$> request
+
+getHeader :: HeaderName -> ActionM Strict.Text
+getHeader name = maybe (raise message) return =<< maybeHeader name
+  where message = "missing \"" <> headerName <> "\" header"
+        headerName = (Lazy.fromStrict . Strict.decodeUtf8 . original) name
+
+route :: Database
+      -> Chan (EmailAddress, Code)
+      -> Chan ResolvedPost
+      -> (ActionM () -> ScottyM ())
+      -> ActionM (Either Response Request)
+      -> ScottyM ()
+route db emailChan socketChan path makeReq = path $ do
+  reqOrRes <- makeReq
+  let dbRes = either return execute reqOrRes
+  res <- liftIO (runReaderT dbRes db)
+  effects <- send res
+  mapM_ (liftIO . perform) effects
```

Comment:
[yes.](https://barnacles.blackfriday/search?utf8=%E2%9C%93&q=mapm&what=all&order=relevance)

> Finally, once we make Applicative a superclass of Monad, mapM is basically never the right combinator for, well, anything. In particular requires far too tight a constraint (Monad rather than Applicative). Generalizing mapM to Applicative, runs afoul of several other issues stated below in the FAQ. On the other hand, going the 'wrong' direction and adding a monomorphized traverse breaks far more code, and makes the scoping problems even worse. Once you bring in a generalized traverse, it seems a very arbitrary choice to avoid bringing in the context necessary to define instances of the abstraction it uses.